### ISSUE TITLE
Options for valid answers to number entry parts

### ISSUE THREAD

**Issue by christianp** [numbas/Numbas#380]
Numbers can be written in various forms, and you might want to restrict the student to only some of them.
The question author should have a list of options for valid input:

* integer
* decimal
* fraction (`a/b`)
* scientific notation. E notation is easiest, but is it worth doing `a*10^b` as well?
* engineering notation (scientific notation, but the exponent must be a multiple of 3, so the coefficient is in the range 1 to 1000)

The decimal places restriction only makes sense for decimal numbers. Sig. figs. applies to everything except fractions.
Scientific notation could have an option to require normalised forms (i.e., things like `300e2` would be invalid)

**Comment by christianp** [comment]
*To do this:*

* Add an option `answerFormat` to [`NumberEntryPart.settings`](https://github.com/numbas/Numbas/blob/master/runtime/scripts/parts/numberentry.js#L99). You'll have to add this option to the corresponding class in `bin/exam.py` as well. The option should be a dictionary mapping format names to true/false values, where true means that the format is allowed. The default should be to allow integers and decimals, but nothing else.
* In the `NumberEntryPart` constructor, change [the definition of `displayAnswer`](https://github.com/numbas/Numbas/blob/master/runtime/scripts/parts/numberentry.js#L60) so it produces an answer in an allowed format.
* Expand `NumberEntryPart.studentAnswerAsFloat` to accept each of the different forms. It should try each of the allowed forms in turn.
* Change `cleanAnswer` so it does the right thing based on the allowed formats.
* Add a method `validAnswer(string)` which decides if a given string is a valid answer, and use it in the `mark` method instead of `util.isNumber`.
* In `themes/default/files/scripts/display.js`, change `NumberEntryPartDisplay.cleanStudentAnswer` to accept all allowed formats, and give sensible feedback when an invalid answer is given.
* Add the option to the editor.

This option supersedes the "allow fractions" option, so you should remove that and write a [.exam migration](https://github.com/numbas/Numbas/blob/master/bin/migrations.py) to upgrade old .exam files.

**Comment by janga1997** [comment]
@christianp I want to take on this issue, but this seems pretty daunting. 
Are all parts of this issue still unresolved?

**Comment by christianp** [comment]
Yes, this is still unresolved, and in fact I want to take it on, so I'll remove the up-for-grabs label.

**Comment by christianp** [comment]
I've added "complex number" to my list of formats, but I'm not completely convinced that it fits. It's somewhere on the line between the single number entry part we currently have and the matrix entry part. I'm not massively happy with how many of the number entry options are repeated for the matrix entry part. And if we allow a "set entry" for things like roots of equations, that would be another part type with many of the same options.

I don't think there's any way round this - I just need to make sure that shared code is not repeated, and accept that options will be repeated across part types. I don't think a massive generic "some number of numbers" part type is the right answer.

**Comment by christianp** [comment]
I've added "number in a range", because we have a few questions on modular arithmetic and trig that ask for a number in a given range. Refusing to accept numbers outside the desired range would help avoid innocent mistakes.

**Comment by christianp** [comment]
The possibility of accepting scientific notation like `2.321 * 10^3` has come up again. I thought I'd write down some thoughts here.

It looks straightforward: you might write down a regular expression like `(-?)(\d+(\.\d+)?)\s*\*\s*10^(-?\d+)`. This would match things like `2*10^4` and `-1.302 * 10^-19`.

But it wouldn't match `—2×10ˆ-2`, or even `-2 * 10^(-2)`, when you use unicode lookalike characters or brackets around the exponent.

So it might make sense to parse the string as a JME expression and then use pattern-matching to check it's of the form ```(`+- $n) * 10^(`+- integer:$n)```. Would we want to accept `10^2 * 2.34`, or should we turn off commutativity? What about `2.34 * 10`?

But then we should really provide a preview rendering to the student. Is that reasonable, even though most of the time it'll look exactly the same as what they wrote?

**Comment by christianp** [comment]
One reason not to use the JME parser for scientific notation is that at the moment there's no way of getting it to stop once it's matched "enough". The other number formats all allow for any amount of extra stuff after the number part, to allow for strings like a quantity with units.

**Comment by christianp** [comment]
We've had a request to use scientific notation for the parts of a complex number, e.g. `1e3 + 3e2i`.

**Comment by christianp** [comment]
Engineers sometimes write complex numbers in the format `a + jb`, e.g. `1 + j2`.
